File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/sessions/PlaySession.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.sessions;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionActor;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher;
import systems.zlink.framework.streams.ZLinkStreamError;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

// --8<-- [start:doc-session]
public final class PlaySession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers;

    public PlaySession(
            ZLinkSessionContext context,
            ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers) {
        this.context = context;
        this.handlers = handlers;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnected() {
        return CompletableFuture.allOf(
                context.actors().bound().stream()
                        .map(actor -> actor.notifyDisconnected().toCompletableFuture())
                        .toArray(CompletableFuture[]::new));
    }

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDispatch(
            ZLinkSessionDispatchContext header, ZLinkMessage payload) {
        return handlers.tryHandle(context, header, payload)
                .thenCompose(
                        handled ->
                                handled
                                        ? CompletableFuture.completedFuture(null)
                                        : requireActor(header.packetName())
                                                .relay(header, payload)
                                                .thenApply(ignored -> null));
    }

    private ZLinkSessionActor requireActor(String packetName) {
        return switch (context.actors().bound().size()) {
            case 1 -> context.actors().bound().get(0);
            case 0 ->
                    throw new IllegalStateException(
                            "AuthenticateReq is required before play packet '" + packetName + "'");
            default ->
                    throw new IllegalStateException(
                            "Exactly one actor must be bound before play packet '"
                                    + packetName
                                    + "'");
        };
    }
}
// --8<-- [end:doc-session]

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/sessions/PlaySession.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.sessions

import kotlinx.coroutines.future.await
import systems.zlink.framework.kotlin.ZLinkSuspendingSession
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSessionActor
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher

// --8<-- [start:doc-session]
class PlaySession(
    private val context: ZLinkSessionContext,
    private val handlers: ZLinkSessionPacketDispatcher<ZLinkSessionContext>,
) : ZLinkSuspendingSession() {
    override fun context(): ZLinkSessionContext = context

    override suspend fun onDisconnectedSuspending() {
        context.actors().bound().forEach { actor -> actor.notifyDisconnected().await() }
    }

    override suspend fun onDispatchSuspending(
        header: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        if (handlers.tryHandle(context, header, payload).await()) {
            return
        }
        requireActor(header.packetName()).relay(header, payload).await()
    }

    private fun requireActor(packetName: String): ZLinkSessionActor =
        when (context.actors().bound().size) {
            1 -> context.actors().bound()[0]
            0 ->
                throw IllegalStateException(
                    "AuthenticateReq is required before play packet '$packetName'"
                )
            else ->
                throw IllegalStateException(
                    "Exactly one actor must be bound before play packet '$packetName'"
                )
        }
}
// --8<-- [end:doc-session]
